### PR TITLE
Create decoder class

### PR DESCRIPTION
**Pull request by erichan1**
Summary: Implement decoder-only layer, which doesn't currently exist in torch. Implement incremental and forced decoding with a new multiheadattention and decoder forward pass. Rather similar to the transformer_encoder_layer_fwd. Not a public facing API, although may become public facing eventually. 

Stacked on top of https://github.com/pytorch/pytorch/pull/79437.

Test Plan: See D36140513 numerical tests.

Reviewed By: mikekgfb

Differential Revision: D36987004

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -478,5 +490,182 @@ std::tuple<Tensor, Tensor> native_multi_head_attention(
...
+  debug_assert_shape(__LINE__, qkt, {B, num_head, T, T});
+#endif
+#ifdef DEBUG_PRINT_EACH_STEP
+  std::cerr << "qkt: " << qkt << std::endl;
+#endif
+
+  // shape: [B, num_head, T, T]
+  // TODO: long-term, have a kernel that works with
+  // NestedTensor directly if there is no mask passed
+  qkt = masked_softmax(qkt, mask, query);
```
Comment (zrphercule) [inline aten/src/ATen/native/transformers/attention.cpp:634]: Do we even have nested masedk softmax already?

**Review by zrphercule — COMMENTED** [review]
Could you plz add some unit test in test/test_nn.py ? Seems these changes are no where tested

File: aten/src/ATen/native/transformers/attention.cpp (hunk 1 above)
Comment (erichan1) [reply aten/src/ATen/native/transformers/attention.cpp:634]: I copied this from transformer_encoder_layer_fwd, so I don't believe so. On testing, would you recommend unit tests that just check for invalid shapes like so https://github.com/pytorch/pytorch/blob/543919cfc8d7014dfa365b55c459fa2b3f746891/test/test_nn.py#L5878? I don't see any numerical testing for transformer_encoder_layer_fwd. 

I have numerical testing in an internal benchmarking script where I compare against fairseq, but putting that test in torch.testing would mean I need to import fairseq in the test, which isn't a reliable dependency. I can't compare against nn.TransformerDecoder because we don't have an existing decoder-only layer, only encoder-decoder.

Comment (erichan1) [reply aten/src/ATen/native/transformers/attention.cpp:635]: Tests are in test_transformers.py here https://github.com/pytorch/pytorch/pull/79576.

**Review by zrphercule — APPROVED** [review]
Great job!
